### ISSUE TITLE
Ocaml generated code uses unsafe Obj.magic regarding flambda

### ISSUE THREAD

**Issue by mjambon** [mjambon/atd#96]
_From @jorisgio on April 28, 2016 11:38_

All code i tested with ocaml 4.03.0+flambda segfault in deserialization functions.

_Copied from original issue: mjambon/atdgen#45_

**Comment by mjambon** [comment]
_From @gasche on May 5, 2016 21:5_

We had one more segfault report from an user today on IRC -- I think they're planning on sending a bug report here. I defensively proposed ( ocaml/opam-repository#6454 ) to mark atdgen `< 4.03` on opam-repository to prevent further issues. I don't know yet whether the segfaults are flambda-only -- maybe both versions are affected.

**Comment by mjambon** [comment]
I haven't looked into this and I'm unlikely to take that time soon. I might just make atdgen's generated code no longer on obj magic at all because it's less maintenance.

**Comment by mjambon** [comment]
_From @gasche on May 5, 2016 21:42_

For the record, I would attempt to guess that it's your use of `Obj.magic 0.0` that is at fault: if the compiler got more clever at specialization, then it can probably tell that you are actually building a float array (looking at `-dcmm` output would tell you). 4.03 introduce a new `Sys.opaque_identity` function that prevents the compiler from making assumption on its value; if you can detect the compiler version you're producing code against, I think it would be wise to use something like `Obj.magic (Sys.opaque_identity None)` instead (`Sys.opaque_identity None` is an `'a option` that may or may not be a pointer).

PS. But yeah, not using `Obj` at all is possibly the best choice.

**Comment by mjambon** [comment]
_From @leowzukw on May 6, 2016 16:14_

I’m the user on IRC, @gasche thanks for the bug report.

**Comment by mjambon** [comment]
_From @jorisgio on May 13, 2016 15:51_

`Obj.magic (Sys.opaque_identity None)` indeed doesn't segfault on the 5 atd files i tried (although i have not had time to implement it in atdgen generator.

**Comment by mjambon** [comment]
_From @gasche on May 13, 2016 17:17_

Very useful, thanks! You should send @mjambon a pull-request, and we could have a bugfix release on OPAM that is not known to work perfectly, but at least doesn't directly segfault, waiting for a better long-term fix.

**Comment by mjambon** [comment]
_From @jvillard on May 17, 2016 12:55_

FYI, there is also a crash in [infer](https://github.com/facebook/infer) with 4.03.0 without flambda which may be related:

``` sh
opam switch 4.03.0
eval $(opam config env)
# get an Infer source release from https://github.com/facebook/infer/releases/tag/v0.8.1
wget https://github.com/facebook/infer/releases/download/v0.8.1/infer-linux64-v0.8.1.tar.xz
tar xf infer-linux64-v0.8.1.tar.xz
cd infer-linux64-v0.8.1
./build-infer.sh clang # crashes at the end of the compilation process, when running Infer on the C/C++ models
```

Here's a more precise repro to run after the steps above:

``` sh
./infer/bin/infer -fd -- clang -c infer/models/c/src/wchar.c
sh wchar.o.sh # crashes
# here is what the small shell script above runs in more details, to access the crashing OCaml process
# generate the biniou file from C++ via clang, this works fine
./infer/lib/clang_wrappers/clang_wrapper -c infer/models/c/src/wchar.c -Xclang -load -Xclang ./infer/lib/clang_wrappers/../../../facebook-clang-plugins/libtooling/build/FacebookClangPlugin.dylib -Xclang -add-plugin -Xclang BiniouASTExporter -Xclang -plugin-arg-BiniouASTExporter -Xclang - -Xclang -plugin-arg-BiniouASTExporter -Xclang PREPEND_CURRENT_DIR=1  > wchar.o.biniou
# read the biniou file in OCaml land, segfault!
./infer/bin/InferClang -c infer/models/c/src/wchar.c -debug -x c -ast wchar.o.biniou
```

**Comment by mjambon** [comment]
_From @gasche on June 11, 2016 22:20_

#46 proposes using `Obj.opaque_identity` as described above.

**Comment by mjambon** [comment]
Fix https://github.com/mjambon/atdgen/pull/46 by @gasche is part of atdgen 1.9.1. I'm making the opam package available for 1.9.1 to all versions of OCaml (https://github.com/ocaml/opam-repository/pull/6799). If there's a still a segfault, let us know.

**Comment by mjambon** [comment]
_From @leowzukw on June 20, 2016 20:20_

Works great [with OcLaunch](https://gitlab.com/WzukW/oclaunch/commit/7208d6b0600c1abfc8ee27ebe8f10fd12caf8a8b), thanks.
